### ISSUE TITLE
node-gyp rebuild hooks

### ISSUE THREAD

**Issue by computerquip-streamlabs** [cjntaylor/node-cmake#36]
It's unclear how to take advantage of rebuild hooks when including a package in node_modules. 
Cmake-js for example, allows you to specify a `cmake-js` field in the top-most package.json. Since ncmake appears to ignore node-gyp environment (such as msvs_version), there doesn't appear to be a reliable way to take advantage of the rebuild hooks.

**Comment by cjntaylor** [comment]
What functionality from the node-gyp environment / `cmake-js` do you need? I'm happy to add support for parsing a `ncmake` field in the `package.json`, but I'm not sure what I'd put in there that isn't already covered by the arguments you can pass to `nodejs_init()` Since these are so specific to the environment you want to run in (node vs electron, etc), I've always felt like it was much clearer to put these in your `CMakeLists.txt`.

`msvs_version` is interesting / sort of odd to me. I'm not quite sure how to handle it, as it seems like it'd be highly prone to breaking modules if the required version isn't installed on the system trying to include the module. Right now, `ncmake` uses the "default" CMake generator, which is selected based on **what is actually installed on the target** by some blackboxed CMake logic. The downside is that if you don't have a version of Visual Studio installed that is compatible with the module/active version of node, it'll also break. But more likely than not, the default is at least likely to be new enough to cause the fewest issues. While its a bit ugly, it is possible to specify the `-g` parameter and select a visual studio version that way:

```bash
ncmake -g "Visual Studio 12 2013" rebuild
```

This isn't much better and will also fail on systems without 2013 installed (and this won't work with recent versions of node/v8 which require VS2015), but it'll work if you need this now. 

None of this feels like a particularly good solution, but I don't feel like adding the `package.json` keys enables anything that isn't already provided by the module. I'm open to suggestions / PRs though :wink:

**Comment by computerquip-streamlabs** [comment]
I'm not sure what I would actually want here when I think harder on it. 

There's two methods native addon's are distributed:
1. They're given in a pre-built form where no compilation is required. Our project recently moved to this since we have very specific needs. 

2. They distribute source and define build steps in the "install" script. When a project then includes the module in their project, it builds it on the fly. This is the point of `yarn rebuild` as well. 

The problem with number one is that until N-API is common place (and even then, you'll be sporting at least 6 different prebuild packages if you support only the three big platforms), you have to rebuild for every combination of operating system, architecture, and ABI change in node. Also note that Electron uses a different Node ABI than upstream so native node modules must be recompiled specially for Electron. 

The problem with number two is that it's slow (sometimes ridiculously so) and actually getting a native addon to compile on the fly, especially one with dependencies, can be tricky. In the case of using cmake, the slowness could be helped if we use Ninja which unfortunately complicates our environment a little bit more for the non-savvy since it requires `cl` or `gcc` be in path for Windows platforms. Also, since builds should be reproducible to begin with, the result will ultimately be the same as what the prebuild will provide. An additional problem is that you have to be able to tell the build system what runtime you want to build against. If your install command looks something like: `cmake -H. -Bbuild && cmake --build build --config Release`, there's nothing you can really do to specify a runtime or version here. You could make a node script that reads off environment variables yourself but the behavior depends on whether you're running yarn, npm, or some other npm replacement, and how they handle the environment of scripts run in sub packages. In my tests, it varies drastically, where yarn didn't work correctly, and npm worked but only sometimes (far to frustrated to figure out why).

In retrospect, ncmake is ultimately two things: the CMake module and the node module for importing the cmake module. You don't even need ncmake the command to get things to work. It would be somewhat difficult to get the cmake module to do what I'm asking. cmake-js does a bit of magic to read from the package.json in the top-most directory and even that isn't perfect all the time ( seems to have issues performing recursively at times and yarn is constantly confused about it ). node-gyp is really the only one that's natively supported by node and npm makes sure that any project that uses node-gyp has the correct disturl, runtime, and version, but pretty much doesn't give two poops about external build systems and from what I've seen, doesn't expose that information in a way you can cleanly fetch from within a node script. 

TL;DR The environment in npm for native addons is horrible. There's no good way of doing this right now, and it's not the fault of the build system. I'm closing this until node-gyp is thrown into the garbage where it belongs and node gets better support for external build systems. I also imagine the adoption of N-API might help ease some of the growing pains.

**Comment by cjntaylor** [comment]
> In retrospect, ncmake is ultimately two things: the CMake module and the node module for importing the cmake module. You don't even need ncmake the command to get things to work. It would be somewhat difficult to get the cmake module to do what I'm asking. cmake-js does a bit of magic to read from the package.json in the top-most directory and even that isn't perfect all the time ( seems to have issues performing recursively at times and yarn is constantly confused about it ). node-gyp is really the only one that's natively supported by node and npm makes sure that any project that uses node-gyp has the correct disturl, runtime, and version, but pretty much doesn't give two poops about external build systems and from what I've seen, doesn't expose that information in a way you can cleanly fetch from within a node script.

I was confused what you meant by this at first as it still doesn't really answer the question I asked, but let me echo back what I think you're saying:

*If a project using `node-cmake` is a dependency of another module, it should build against the same disturl,runtime and version as the top level module in the chain. This is apparently handled by `node-gyp` and badly by `cmake-js`* 

(how did you find this out? How does npm pass this information to node-gyp?)

If this is what you mean, I agree, but I likewise have no idea right now how to handle this. In theory, this information could be passed right now to the current cmake script via the environment variables `NODEJS_URL`, `NODEJS_NAME` and `NODEJS_VERSION`, which will already override the defaults (in fact, this is how some of the flags you can pass to `ncmake` are implemented), but like you said, I'm not sure how I'd handle this as you'd walk down the module chain (or if npm even sets/preserves ENV as it walks the dependency set) 🤔 

Totally understand on the need for pre-built. I'm assuming you're linking statically to any dependencies so that you don't have to rely on a system having missing native deps? How do you handle the possible symbol collision (e.g. If your module statically links against openssl, all of those symbols will collide with the ones statically linked into the node binary)? I've had this on my list of things to think about for this for as long as I can remember and its remained tabled because "it's complicated".

Also agreed on the state of native addons; its in a terrible state of flux right now due to the pickle the node development team is in with google mostly dropping support for using gyp to build v8 (and doing it again switching from `gn` to `bazel`), and picking a non-open replacement (this is totally valid from their end, but now node has to either opt to maintain gyp or find a replacment, and thats been taking forever). I've had some limited interaction with the folks who are supposedly working on phasing out gyp support but they haven't talked to me in quite some time (I think before the 2.0 rewrite), so I don't really know whats going on at this point.

N-API is news to me but its been needed for a while; hopefully that can be solidified and we can stop needing NAN and a bunch of really esoteric knowledge of v8 internals to properly write addons. But it'll take a while.

When I can, I'll poke around the environment variable stuff / try to figure out how this is handled by npm but no promises. If you think of anything I can add to `NodeJS.cmake` or `ncmake` that would help with this or your prebuild process, I'm happy to make changes / accept PRs.

Thanks for the feedback :smile:

**Comment by computerquip-streamlabs** [comment]
I statically link nothing. I don't really gain a whole lot by statically linking. I distribute the required dependencies with the module itself.
